scalding-parquet-scrooge/src/main/scala/com/twitter/scalding/parquet/scrooge/ParquetListFormatRule.scala

```diff
@@ -0,0 +1,262 @@
+    // "the repeated field is not a group,
+    // then its type is the element type and elements are required."
+    true
+  }
+
+  override def check(repeatedType: Type): Boolean =
+    repeatedType.isPrimitive && repeatedType.getName == this.constantElementName
+
+  override def createCompliantRepeatedType(typ: Type, name: String, isElementRequired: Boolean, originalType: OriginalType): Type = {
+    if (!isElementRequired) throw new IllegalArgumentException("Primitive list format can only take required element")
+    if (!typ.isPrimitive) throw new IllegalArgumentException(String.format("Primitive list format cannot take group, but is given %s", typ))
+    new PrimitiveType(Type.Repetition.REPEATED, typ.asPrimitiveType.getPrimitiveTypeName, this.constantElementName, originalType)
+  }
+}
+
+private[scrooge] object PrimitiveElementRule extends PrimitiveListRule {
+  override def constantElementName: String = "element"
+}
+
+private[scrooge] object PrimitiveArrayRule extends PrimitiveListRule {
+  override def constantElementName: String = "array"
+}
+
+private[scrooge] sealed trait GroupListRule extends ParquetListFormatRule {
+  /**
+   * repeated group [element|array] {
+   *   required binary str (UTF8);
+   *   required int32 num;
+   * }
+   */
+  def constantElementName: String
+
+  override def isElementRequired(repeatedType: Type): Boolean = {
+    // According Rule 2 from
+    // https://github.com/apache/parquet-format/blob/master/LogicalTypes.md#backward-compatibility-rules
+    // "If the repeated field is a group with multiple fields,
+    // then its type is the element type and elements are required."
+    true
+  }
+
+  override def elementType(repeatedType: Type): Type = repeatedType
+
+  override def elementName(repeatedType: Type): String = this.constantElementName
+
+  override def check(repeatedType: Type): Boolean = {
+    if (repeatedType.isPrimitive) false
+    else {
+      val groupType = repeatedType.asGroupType
+      groupType.getFields.size > 0 && groupType.getName == this.constantElementName
+    }
+  }
+
+  override def createCompliantRepeatedType(typ: Type, name: String, isElementRequired: Boolean, originalType: OriginalType): Type = {
+    if (typ.isPrimitive) new GroupType(Type.Repetition.REPEATED, this.constantElementName, typ)
+    else new GroupType(Type.Repetition.REPEATED, this.constantElementName, typ.asGroupType.getFields)
+  }
+}
+
+private[scrooge] object GroupElementRule extends GroupListRule {
+  override def constantElementName: String = "element"
+}
+
+private[scrooge] object GroupArrayRule extends GroupListRule {
+  override def constantElementName: String = "array"
+}
+
+private[scrooge] object TupleRule extends ParquetListFormatRule {
+  override def check(repeatedType: Type): Boolean = repeatedType.getName.endsWith("_tuple")
+
+  override def elementName(repeatedType: Type): String = {
+    repeatedType.getName.substring(0, repeatedType.getName.length - 6)
```

Comment:
nit: would like docs or variable/method renaming to explain this substring indexing a little more clearly

Reply:
I now made `_tuple` as a variable where length is calculated from it more explicitly.